File: Data Structures & Algorithms/subsets/submission-2.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []

        def backtrack(curr: List[int], i: int) -> None:
            if i == len(nums):
                res.append(curr.copy())
                return
            
            curr.append(nums[i])
            backtrack(curr, i + 1)

            curr.pop()
            backtrack(curr, i + 1)

        backtrack([], 0)
        return res
        

File: Data Structures & Algorithms/permutations/submission-2.java
class Solution {
    
    List<List<Integer>> res;

    public List<List<Integer>> permute(int[] nums) {
        res = new ArrayList<>();
        List<Integer> curr = new ArrayList<>();

        this.backtrack(curr, nums);
        return res;
    }

    public void backtrack(List<Integer> curr, int[] nums) {
        if (curr.size() == nums.length) {
            res.add(new ArrayList<>(curr));
            return;
        }

        for (int i = 0; i < nums.length; i ++) {
            int num = nums[i];

            if (!curr.contains(num)) {
                curr.add(num);
                this.backtrack(curr, nums);
                curr.remove(curr.size() - 1);
            }
        }
    }
}
